walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,64 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    ArrayList<TimeRange> result = new ArrayList<>();
+    List<TimePoint> timePoints = new ArrayList<>();
+    Set<String> attendees = new HashSet<>(request.getAttendees());
+    Set<String> allAttendees = new HashSet<>(request.getOptionalAttendees());
+    long duration = request.getDuration();
+    allAttendees.addAll(attendees);
+    result = queryByAttendenceAndDuration(events, allAttendees, duration);
+    if (result.size() == 0) {
+      result = queryByAttendenceAndDuration(events, attendees, duration);
+    }
+    return result;
+  }
+
+  public ArrayList<TimeRange> queryByAttendenceAndDuration(
+      Collection<Event> events, Set<String> attendees, long duration) {
+    ArrayList<TimeRange> result = new ArrayList<>();
+    List<TimePoint> timePoints = new ArrayList<>();
+    for (Event event : events) {
+      Set<String> curAttendees = new HashSet<>(event.getAttendees());
+      curAttendees.retainAll(attendees);
+      if (!curAttendees.isEmpty()) {
+        TimePoint startTime = new TimePoint(event.getWhen().start(), "start");
+        TimePoint endTime = new TimePoint(event.getWhen().end(), "end");
+        timePoints.add(startTime);
+        timePoints.add(endTime);
+      }
+    }
+    Collections.sort(timePoints, TimePoint.ORDER_BY_TIME);
+    int meetsNum = 0;
+    int prev = TimePoint.START_OF_DAY;
+    boolean isEmpty = true;
```

Comment:
I think these could be slightly clearer too. E.g. `numOngoingMeetings` instead of meetsNum, `previousStartOrEnd` instead of `prev`, and `currentTimeIsAvailable` instead of `isEmtpy`.

These are just suggestions, feel free to pick what you think is most descriptive of what the variable represents.